File: tv/apple-tv/PlayBridge TV/PlayBridge TV/QualityRanker.swift
import Foundation

/// Quality tier rank — mirrors the phone's QualityFilter patterns so both sides
/// apply the same preference. Higher rank = preferred.
struct QualityRanker {
    private static let uhdPatterns = ["2160p", "4k", "uhd"]
    private static let fhdPatterns = ["1080p", "1080"]
    private static let hdPatterns = ["720p", "720"]

    /** Returns 4, 3, 2, or 1 for UHD / FHD / HD / SD respectively based on name/title text. */
    static func rankFromText(_ text: String) -> Int {
        let lower = text.lowercased()
        if uhdPatterns.contains(where: { lower.contains($0) }) { return 4 }
        if fhdPatterns.contains(where: { lower.contains($0) }) { return 3 }
        if hdPatterns.contains(where: { lower.contains($0) }) { return 2 }
        return 1
    }

    /**
     * Returns the target rank for the given quality preference key.
     * Returns 0 for null / "auto" so callers can detect "no preference".
     */
    static func targetRank(_ qualityPreference: String?) -> Int {
        guard let pref = qualityPreference?.lowercased() else { return 0 }
        if uhdPatterns.contains(pref) { return 4 }
        if fhdPatterns.contains(pref) { return 3 }
        if hdPatterns.contains(pref) { return 2 }
        return 0
    }
}

File: shared/src/commonMain/kotlin/com/playbridge/shared/stremio/QualityRanker.kt
package com.playbridge.shared.stremio

/**
 * Quality tier rank — mirrors the phone's QualityFilter patterns so both sides
 * apply the same preference. Higher rank = preferred.
 *
 * Key values match PlayPayload.defaultVideoQuality strings:
 *   "2160p", "1080p", "720p", or null / anything else → best available.
 */
object QualityRanker {

    private val UHD_PATTERNS  = listOf("2160p", "4k", "uhd")
    private val FHD_PATTERNS  = listOf("1080p", "1080")
    private val HD_PATTERNS   = listOf("720p", "720")

    /** Returns 4, 3, 2, or 1 for UHD / FHD / HD / SD respectively based on name/title text. */
    fun rankFromText(text: String): Int {
        // Strip zero-width / format / bidi obfuscation the same way SourceTypeRanker does,
        // so titles like "1‍0‍8‍0‍p" still classify correctly.
        val lower = SourceTypeRanker.normalizeForMatching(text)
        return when {
            UHD_PATTERNS.any { lower.contains(it) } -> 4
            FHD_PATTERNS.any { lower.contains(it) } -> 3
            HD_PATTERNS.any  { lower.contains(it) } -> 2
            else                                    -> 1
        }
    }

    /**
     * Returns the target rank for the given quality preference key.
     * Returns 0 for null / "auto" so callers can detect "no preference".
     */
    fun targetRank(qualityPreference: String?): Int = when (qualityPreference?.lowercase()) {
        "2160p", "4k", "uhd" -> 4
        "1080p", "1080"      -> 3
        "720p",  "720"       -> 2
        else                 -> 0 // null / "auto" → no preference
    }
}
